src/optimizer/plan_generator.cpp

```diff
@@ -1063,4 +1065,90 @@ void PlanGenerator::Visit(const Analyze *analyze) {
                      .Build();
 }
 
+///////////////////////////////////////////////////////////////////////////////
+// CTE
+///////////////////////////////////////////////////////////////////////////////
+
+void PlanGenerator::Visit(const CteScan *cte_scan) {
+  // CteScan has the same output schema as the child plan!
+  NOISEPAGE_ASSERT(children_plans_.size() <= 2, "CteScan needs at most 2 child plans");
+  auto predicate = parser::ExpressionUtil::JoinAnnotatedExprs(cte_scan->GetScanPredicate()).release();
+  RegisterPointerCleanup<parser::AbstractExpression>(predicate, true, true);
+  if (!children_plans_.empty()) {
+    output_plan_ = std::move(children_plans_[0]);
+    // CteScan OutputSchema does not add/drop columns. All output columns of CteScan
+    // are the same as the output columns of the child plan. As such, the OutputSchema
+    // of a CteScan has the same columns vector as the child OutputSchema, with only
+    // DerivedValueExpressions
+    auto idx = 0;
+    auto &child_plan_cols = output_plan_->GetOutputSchema()->GetColumns();
+    std::vector<planner::OutputSchema::Column> child_columns;
+    // A copy of child_columns maintained to create an identical output schema to be
+    // passed on to a child of the current CTE node (the recursive read in recursive/
+    // iterative CTEs)
+    std::vector<planner::OutputSchema::Column> inner_columns;
+    for (auto &col : child_plan_cols) {
+      auto dve = std::make_unique<parser::DerivedValueExpression>(col.GetType(), 0, idx);
+      // Made a second version to avoid ownership issues
+      auto dve_copy = std::make_unique<parser::DerivedValueExpression>(col.GetType(), 0, idx);
+      child_columns.emplace_back(col.GetName(), col.GetType(), std::move(dve));
+      inner_columns.emplace_back(col.GetName(), col.GetType(), std::move(dve_copy));
+      idx++;
+    }
+
+    std::vector<planner::OutputSchema::Column> columns;
+    for (auto &output_expr : output_cols_) {
+      auto tve = output_expr.CastManagedPointerTo<parser::ColumnValueExpression>();
+
+      // output schema of a plan node is literally the columns of the table.
+      // there is no such thing as an intermediate column here!
+      columns.emplace_back(tve->GetColumnName(), tve->GetReturnValueType(), tve->Copy());
+    }
+
+    auto cte_scan_out = std::make_unique<planner::OutputSchema>(std::move(child_columns));
+    (void)cte_scan_out;
+    if (children_plans_.size() == 2) {
```

Comment:
Any asserts we can make here? What about 1 child?

Reply:
We have an assert at the top of this function that `children_plans_` has size 0,1 or 2, and we are inside the `if` statement for the case where it is non-empty, so we do have a guarantee that the the `else` case is for when there is only one child. So I don't think we need an additional assert here.